packages/client/src/features/viewport/viewport-modules.ts

```diff
@@ -30,17 +30,29 @@ import {
 } from '~glsp-sprotty';
 import { GLSPScrollMouseListener } from './glsp-scroll-mouse-listener';
 
-export const viewportModule = new ContainerModule((bind, _unbind, isBound) => {
+export const viewportModule = new FeatureModule((bind, _unbind, isBound) => {
     const context = { bind, isBound };
     configureCommand(context, CenterCommand);
     configureCommand(context, FitToScreenCommand);
     configureCommand(context, GetViewportCommand);
     configureCommand(context, SetViewportCommand);
 
-    bindAsService(context, TYPES.KeyListener, CenterKeyboardListener);
     bindAsService(context, TYPES.MouseListener, ZoomMouseListener);
     bindAsService(context, TYPES.MouseListener, GLSPScrollMouseListener);
 
     configureActionHandler(context, EnableToolsAction.KIND, GLSPScrollMouseListener);
     configureActionHandler(context, EnableDefaultToolsAction.KIND, GLSPScrollMouseListener);
 });
+
+/**
+ * Feature module that is intended for the standalone deployment of GLSP (i.e. plain webapp)
+ * When integrated into an application frame (e.g Theia/VS Code) this module is typically omitted and/or replaced
+ * with an application native module.
+ */
+export const standaloneViewportModule = new FeatureModule(
+    (bind, _unbind, isBound) => {
+        const context = { bind, isBound };
+        bindAsService(context, TYPES.KeyListener, CenterKeyboardListener);
+    },
+    { requires: viewportModule }
```

Comment:
A general thought: why don't we automatically bind the required module (`viewportModule`) if it hasn't been bound already? Maybe this adds too much magic, but on the other hand that could be also what someone may expect when they add a `requires` somewhere.

Reply:
Mhh, this would break the current intended behavior for default modules + application specific modules.
If a default module is missing we have to assume that the adopter intentionally removed it during the container initialization.
In that case we simply don't want to load and modules that require/extend behavior of the removed default module.
Loading the required module in this would magically revert the adopters intentional removal.

The main intended use case (at least from my point of view) is conditional module loading.

For instance if a user removes the `copyPasteModule` because it should not be supported in their Diagram Editor than we also don't want to load any c&p related addtions like context menu entries or keyboard short cuts.
This is also the reason why `failsOnMissingRequired` is false per default.

Maybe we could rework the API a little bit? e.g:
```ts
/**
 * Optional constructor options for {@link FeatureModule}s.
 */
export interface FeatureModuleOptions {
   loadIfPresent?: MaybeArray<FeatureModule>; // set of module that need to be loaded as precondition for this module to load. Does not throw an error or add any of the prerequistes
   requires: MaybeArray<FeatureModule> // set of hard required modules. Will throw an error if the requirements are not ment.
    featureId?: symbol;
}
`
Wdyt?

Reply:
Yes, I see what you mean. I also would like to avoid making this more complex than it needs to be.
From an adopter's point of view, I see two values in this system:
1. Avoiding to load dependents when a module isn't loaded or unloaded.
2. Throw an error if a user tries to load a module for which the dependencies aren't available.

Would you agree? Is there something I miss?
If it is exactly those two, we may want to optimize for them, because currently we don't support both of them at the same time (because `failsOnMissingRequired` is either `true` or `false`), whereas we want to just unload for use case 1 and fail for use case 2. And I'm wondering when `failsOnMissingRequired` would ever be very useful as soon as someone unloads a module purposefully, if this module has dependencies (where they don't expect an error, but they will always retrieve one).
Would it help if we have different behavior for an explicit `unload` call (use case 1), where we don't throw any errors but just unload all dependents and an explicit load call where we throw an error if any of the explicitly mentioned modules have missing dependencies?

I'd also be fine if we first address use case 1 with your original proposal and don't put use case 2 in the focus at all. For some reason, I was mainly thinking about that use case as I read `failOnMissingRequired` and remember cases where adopters ran into issues, because they loaded a module without knowing that they need to load a dependency too.

Reply:
>Yes, I see what you mean. I also would like to avoid making this more complex than it needs to be.
>From an adopter's point of view, I see two values in this system:

> 1. Avoiding to load dependents when a module isn't loaded or unloaded.
> 2. Throw an error if a user tries to load a module for which the dependencies aren't available.

Agreed, those are the values of the system.

>If it is exactly those two, we may want to optimize for them, because currently we don't support both of them at the same time (because failsOnMissingRequired is either true or false), whereas we want to just unload for use case 1 and fail for use case 2. And I'm wondering when failsOnMissingRequired would ever be very useful as soon as someone unloads a module purposefully, if this module has dependencies (where they don't expect an error, but they will always retrieve one).

I would argue that a mixed case isn't very useful anyways. What are the usecases that would require the module to not load for one dependent module and fail for another? And how would we resolve conflicting cases i.e. when dependencies are missing that require both not loading and failing?


We could also argue, that the `fail on required module` case is already handled decently enough by inversify. (missing binding error which indicates that a module is missing) and just cover the `not load on missing dependency usecase´ for now.

>I'd also be fine if we first address use case 1 with your original proposal and don't put use case 2 in the focus at all. For some reason, I was mainly thinking about that use case as I read failOnMissingRequired and remember cases where adopters ran into issues, because they loaded a module without knowing that they need to load a dependency too.

I see. Maybe we could also mitigate this issue by catching the error throw by inversify and rethrow/print a more clear error message?

Reply:
> I see. Maybe we could also mitigate this issue by catching the error throw by inversify and rethrow/print a more clear error message?

Yes that could possibly help! But we can also move this out of this PR imho (even after the Release if necessary). As you prefer and depending on how much additional effort that would case.

Thanks!